src/iframe-helper.js

```diff
@@ -17,6 +17,155 @@
 import {dev} from './log';
 import {parseUrl} from './url';
 
+/**
+ * @typedef {{
+ *   frame: !Element,
+ *   events: !Object<string, !Array<function(!Object)>>
+ * }}
+ */
+let WindowEventsDef;
+
+/**
+ * Returns a mapping from a URL's origin to an array of windows and their listenFor listeners.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {boolean} opt_create create the mapping if it does not exist
+ * @return {?Object<string, !Array<!WindowEventsDef>>}
+ */
+function getListenFors(parentWin, opt_create) {
+  let listeningFors = parentWin.listeningFors;
+
+  if (!listeningFors && opt_create) {
+    listeningFors = parentWin.listeningFors = Object.create(null);
+  }
+  return listeningFors || null;
+}
+
+/**
+ * Returns an array of WindowEventsDef that have had any listenFor listeners registered for this origin.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {string} origin the child window's origin
+ * @param {boolean} opt_create create the array if it does not exist
+ * @return {?Array<!WindowEventsDef>}
+ */
+function getListenForOrigin(parentWin, origin, opt_create) {
+  const listeningFors = getListenFors(parentWin, opt_create);
+  if (!listeningFors) {
+    return listeningFors;
+  }
+
+  let listenOrigin = listeningFors[origin];
+  if (!listenOrigin && opt_create) {
+    listenOrigin = listeningFors[origin] = [];
+  }
+  return listenOrigin || null;
+}
+
+/**
+ * Returns an mapping of event names to listenFor listeners.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {!Element} iframe the iframe element who's context will trigger the
+ *     event
+ * @return {?Object<string, !Array<function(!object)>>}
+ */
+function getOrCreateListenForEvents(parentWin, iframe) {
+  const origin = parseUrl(iframe.src).origin;
+  const listenOrigin = getListenForOrigin(parentWin, origin, true);
+
+  let windowEvents = listenOrigin.find(we => {
+    return we.frame === iframe;
+  });
+
+  if (!windowEvents) {
+    windowEvents = {
+      frame: iframe,
+      events: Object.create(null),
+    };
+    listenOrigin.push(windowEvents);
+  }
+
+  return windowEvents.events;
+}
+
+/**
+ * Returns an mapping of event names to listenFor listeners.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {string} origin the child window's origin
+ * @param {!Window} triggerWin the window that triggered the event
+ * @return {?Object<string, !Array<function(!object)>>}
+ */
+function getListenForEvents(parentWin, origin, triggerWin) {
+  const listenOrigin = getListenForOrigin(parentWin, origin);
+
+  if (!listenOrigin) {
+    return listenOrigin;
+  }
+  const windowEvents = listenOrigin.find(we => {
```

Comment:
In busy code it may make sense to write this using a `for` loop to avoid allocating the closure.